### PR TITLE
Logs delegation sc meta

### PR DESCRIPTION
**Pull request by miiu96**
Added logs and events for the delegation system smart contract.

### CODE REVIEW

File: vm/mock/systemEIStub.go [hunk 1]
```diff
@@ -39,6 +39,9 @@ type SystemEIStub struct {
 	ReturnMessage                       string
 }
 
+func (s *SystemEIStub) AddLogEntry(_ *vmcommon.LogEntry) {
```
Comment (sasurobert) [inline vm/mock/systemEIStub.go:42]: add log

File: vm/systemSmartContracts/delegation.go [hunk 2]
```diff
@@ -797,6 +821,13 @@ func (d *delegation) setAutomaticActivation(args *vmcommon.ContractCallInput) vm
 		return vmcommon.UserError
 	}
 
+	entry := &vmcommon.LogEntry{
```
Comment (sasurobert) [inline vm/systemSmartContracts/delegation.go:824]: make a new function which creates this entry and adds it. Thus createAndAdd - you may get the args as vmcommon.ContractCallInput as input, plus [][]byte as topics

Comment (sasurobert) [reply vm/systemSmartContracts/delegation.go:824]: and if input args is nil, maybe use args.Arguments.

File: vm/systemSmartContracts/delegation.go [hunk 3]
```diff
@@ -1140,6 +1206,13 @@ func (d *delegation) unStakeNodes(args *vmcommon.ContractCallInput) vmcommon.Ret
...
```
Comment (sasurobert) [inline vm/systemSmartContracts/delegation.go:1209]: I do not see why should we add this entry - this does not have any value. The same for everywhere you add directly hex decoded data as logEntry.

File: vm/systemSmartContracts/delegation.go [hunk 4]
```diff
@@ -1680,6 +1776,15 @@ func (d *delegation) unDelegate(args *vmcommon.ContractCallInput) vmcommon.Retur
...
+		Identifier: []byte(args.Function),
+		Address:    args.CallerAddr,
+		Topics: [][]byte{
+			valueToUnDelegate.Bytes(), delegator.ActiveFund, globalFund.TotalActive.Bytes(),
```
Comment (sasurobert) [inline vm/systemSmartContracts/delegation.go:1783]: delegator.ActiveFund is only a key - it has no relevant information. You need the actual value as big.Int.bytes to return here

File: vm/systemSmartContracts/delegation.go [hunk 5]
```diff
@@ -2052,6 +2164,22 @@ func (d *delegation) withdraw(args *vmcommon.ContractCallInput) vmcommon.ReturnC
...
+		d.eei.AddReturnMessage(errGet.Error())
+		return vmcommon.UserError
+	}
+
+	numUsers := big.NewInt(0).SetUint64(dStatus.NumUsers)
+	entry := &vmcommon.LogEntry{
+		Identifier: []byte(args.Function),
+		Address:    args.CallerAddr,
+		Topics: [][]byte{
+			totalUnBondable.Bytes(), delegator.ActiveFund, numUsers.Bytes(), globalFund.TotalActive.Bytes(),
```
Comment (sasurobert) [inline vm/systemSmartContracts/delegation.go:2178]: use getFund(delegator.ActiveFund).Bytes()

File: vm/systemSmartContracts/logs.go [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+	numUsersWithCurrent := dStatus.NumUsers
+	if isNew {
+		numUsersWithCurrent++
+	}
+
+	activeFund := big.NewInt(0)
+	if len(delegator.ActiveFund) != 0 {
+		fund, err := d.getFund(delegator.ActiveFund)
+		if err != nil {
+			log.Warn("d.createLogEntryForDelegate cannot get fund", "error", err.Error())
```
Comment (sasurobert) [inline vm/systemSmartContracts/logs.go:27]: this is not log.Warn - this is critical error. return it as.

File: vm/systemSmartContracts/logs.go [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+			activeFund = fund.Value
+		}
+	}
+
+	numUsers := big.NewInt(0).SetUint64(numUsersWithCurrent)
+	numActiveWithCurrentValue := big.NewInt(0).Add(globalFund.TotalActive, delegationValue)
+	delegatorActiveWithCurrent := big.NewInt(0).Add(activeFund, delegationValue)
+	return &vmcommon.LogEntry{
+		Identifier: []byte(funcIdentifier),
+		Address:    callerAddr,
```
Comment (sasurobert) [inline vm/systemSmartContracts/logs.go:38]: indent this and topics as well

**Review by sasurobert — CHANGES_REQUESTED** [review]

File: vm/systemSmartContracts/delegation.go (hunk 2 above)
Comment (miiu96) [reply vm/systemSmartContracts/delegation.go:824]: done.

File: vm/systemSmartContracts/delegation.go (hunk 3 above)
Comment (miiu96) [reply vm/systemSmartContracts/delegation.go:1209]: removed.

File: vm/systemSmartContracts/delegation.go (hunk 4 above)
Comment (miiu96) [reply vm/systemSmartContracts/delegation.go:1783]: fixed now use `remainedFund.Bytes()`

Comment (miiu96) [reply vm/systemSmartContracts/delegation.go:1783]: refactored.

File: vm/systemSmartContracts/delegation.go (hunk 5 above)
Comment (miiu96) [reply vm/systemSmartContracts/delegation.go:2178]: refactored.

File: vm/systemSmartContracts/logs.go (hunk 7 above)
Comment (miiu96) [reply vm/systemSmartContracts/logs.go:38]: done.

File: vm/systemSmartContracts/logs.go (hunk 6 above)
Comment (miiu96) [reply vm/systemSmartContracts/logs.go:27]: refactored this part and kept the log.

File: vm/systemSmartContracts/delegation.go [hunk 8]
```diff
@@ -571,7 +571,14 @@ func (d *delegation) mergeValidatorDataToDelegation(args *vmcommon.ContractCallI
 		return vmcommon.UserError
 	}
 
-	return d.delegateUser(validatorData.TotalStakeValue, big.NewInt(0), validatorAddress, args.RecipientAddr, dStatus)
+	entry := &vmcommon.LogEntry{
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/delegation.go:574]: please replace all these (12) occurrences with a call to `createAndAddLogEntry`

**Review by iulianpascalau — CHANGES_REQUESTED** [review]

File: vm/systemSmartContracts/eei.go [hunk 9]
```diff
@@ -28,6 +28,8 @@ type vmContext struct {
 
 	returnMessage string
 	output        [][]byte
+
```
[the remark this answers is not in the document]
Comment (miiu96) [reply vm/systemSmartContracts/eei.go:31]: done.

File: vm/systemSmartContracts/delegation.go (hunk 3 above)
Comment (miiu96) [reply vm/systemSmartContracts/delegation.go]: done.

File: vm/systemSmartContracts/delegation.go (hunk 8 above)
Comment (miiu96) [reply vm/systemSmartContracts/delegation.go:574]: done.

**Review by sasurobert — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Pull request merged**